servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/MpmcSequentialRunQueue.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.netty;
+
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import javax.annotation.Nullable;
+
+import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
+
+/**
+ * This queue allows for Muli-Producer Multi-Consumer (Mpmc) threading semantics while also invoking the head
+ * {@link Node}'s {@link Node#run()} in a serial fashion. The {@link Node#run()} will eventually trigger a
+ * {@link #poll(Node)} (possibly asynchronously on another thread) which will invoke {@link Node#run()} on the next head
+ * (assuming one exists).
+ * <p>
+ * Below is the expected interaction pattern and lifecycle of a Node:
+ *     <pre>{@link #offer(Node)} -> {@link Node#run()} -> {@link #poll(Node)}</pre>
+ * <p>
+ * Although this queue supports Multi-Consumer threading semantics the {@link #poll(Node)} is typically only invoked
+ * from a single thread (assuming successful runnable completion), it maybe invoked multiple times (potentially from
+ * different threads) with the same {@link Node} due to cancellation/failure.
+ */
+final class MpmcSequentialRunQueue {
+    private static final AtomicReferenceFieldUpdater<MpmcSequentialRunQueue, Node>
+            tailUpdater = newUpdater(MpmcSequentialRunQueue.class, Node.class, "tail");
+
+    @Nullable
+    private volatile Node tail;
+
+    /**
+     * Offer {@link Node} to this queue.
+     * @param node The {@link Node} to append. Must only exclusively be offered to this queue instance.
+     */
+    void offer(final Node node) {
```

Comment:
nit: do we need to guard against `null` ? As otherwise we could end up updating the `tail` with null and then fail once we try to run `node.run()` (if node is null). I think at least we should add an `assert`.